File: Quaternion.py
#Quaternion Class in Python.
import math
import sys
class Vector3:
    def __init__(self,x_comp=0,y_comp=0,z_comp=0) -> None:
        """[Initialize Vector Components]"""
        self._x=x_comp;self._y=y_comp;self._z=z_comp

    def Magnitude(self):
        """[Returns the magnitude of the vector]"""
        mag=math.sqrt((self._x**2)+(self._y**2)+(self._z**2))
        return round(mag,Quaternion.prec)

    def Conjugate(self):
        """[Returns the conjugate of the Vector]"""
        return Vector3(-1*self._x,-1*self._y,-1*self._z)    
    
    def ToUnitVector(self):
        """[Return a unit vector(Direction) from the current vector object]"""
        return self/(self.Magnitude())

    def __add__(self,other):
        """[Add two vectors and return their result(new vector)]"""
        return Vector3((self._x+other._x),(self._y+other._y),(self._z+other._z))

    def __sub__(self,other):
        """[Subtract two vectors and return their result(new vector)]"""
        return Vector3((self._x-other._x),(self._y-other._y),(self._z-other._z))

    def __mul__(self,factor):
        """[Scaler multiply two vectors and return their result(new vector)]"""
        return Vector3((self._x*factor),(self._y*factor),(self._z*factor))

    def __truediv__(self,factor):
        """[Scaler divide two vectors and return their result(new vector)]"""
        try:
            v1=self._x/factor;v2=self._y/factor;v3=self._z/factor
            return Vector3(round(v1,Quaternion.prec),round(v2,Quaternion.prec),round(v3,Quaternion.prec))
        except ZeroDivisionError as _:
            print("Division by Zero not possible")
            sys.exit("[ExitInfo]:ZeroDivisionError")
      
    def __repr__(self) -> str:
        """[Represent vector object]"""
        return "Vector3({},{},{})".format(self._x,self._y,self._z)

class Quaternion:
    prec=4
    def __init__(self,scalerPart=0,vector3Comp=Vector3()) -> None:
        """[Intialize Quaternion inputs]"""
        self._a=scalerPart; self._vector=vector3Comp

    def DotProduct(self,other:"Quaternion")->float:
        """[returns the dot product of self with other Quaternion]"""
        return (self._a*other._a)+(self._vector._x*other._vector._x)+(self._vector._y*other._vector._y)+(self._vector._z*other._vector._z)

    def AngularDifference(self,other:"Quaternion")->float:
        """[returns the angular difference(in radians) between two Quaternions]"""
        return math.acos(self.DotProduct(other)/(self.Norm()*other.Norm()))
    
    def isUnitQuaternion(self)->bool:
        """[Check wheather a quaternion has unit norm or not]"""
        return math.isclose(self.Norm(),1.0)

    def Versor(self)->"Quaternion":
        """[Returns a unit Quaternion from self(Quaternion Normalization)]"""
        if math.isclose(self.Norm(),0.0):
            print("Versor does not exist for the Quaternion:{}".format(self))
            sys.exit("[INPUT ERROR]Exit_Message:ZeroNormError")

        sPart= round((self._a/self.Norm()),Quaternion.prec); vect1=self._vector/self.Norm()
        return Quaternion(sPart,vect1)
    
    def Reciprocal(self)->"Quaternion":
        """[Returns the reciprocal of the current Quaternion]"""
        return self.Conjugate()/(self.Norm()**2) 
    
    def Norm(self)->float:
        """[Returns the Norm(Magnitude) of the Quaternion]"""
        norm=math.sqrt(((self._a)**2)+((self._vector._x)**2)+((self._vector._y)**2)+((self._vector._z)**2))
        return round(norm,Quaternion.prec)

    def vectorPart(self):
        """[Returns the tuple of vector part of the Quaternion]"""
        return (self._vector)

    def scalerPart(self):
        """[Returns tuple containing scaler part of the Quaternion]"""    
        return (self._a,)
    
    def InverseQuaternion(self):
        """[returns the inverse quaternion with respect to self]"""
        return self.Conjugate()/(self.Norm()**2)
    
    def Conjugate(self):
        """[Returns the Conjugate of the Quaternion]"""        
        return Quaternion(self._a,self._vector.Conjugate())
    
    def isScaler(self)->bool:
        """[Check wheather the Quaternion is scaler(real) or not]"""    
        return self._vector._x==0 and self._vector._y==0 and self._vector._z==0 
        
    def isPure(self):
        """[Check wheather the Quaternion is pure(scaler term 0) or not]"""
        return self._a==0
    
    def __add__(self,other:"Quaternion")->"Quaternion":
        """[Add two Quaternions and return their result]"""
        if isinstance(other,Quaternion):
            sPart=self._a+other._a
            vPart=self._vector+other._vector
            print(self._vector,other._vector,vPart)
            return Quaternion(sPart,vPart)
        else:
            raise Exception("Incompatible type(s) for addition with Quaterion")

    def __sub__(self,other:"Quaternion")->"Quaternion":
        """[Subtract two Quaternions and return their result]"""
        if isinstance(other,Quaternion):
            return Quaternion(round(self._a-other._a),(self._vector-other._vector))
        else:
            raise Exception("Incompatible type(s) for subtraction with Quaterion")

    def __mul__(self,other:"Quaternion")->"Quaternion":
        #For scaler multiplication
        if isinstance(other,int) or isinstance(other,float) :
            return Quaternion((self._a*other),self._vector*other)
        
        #For Quaternion multiplication: (i**2= j**2 = k**2 = ijk = -1, ij=k ,jk=i ,ki=j ,ji= -k ,kj= -i ,ik= -j ) 
        elif isinstance(other,Quaternion):
            scalerPart= round((self._a*other._a) -(self._vector._x*other._vector._x) - (self._vector._y*other._vector._y) - (self._vector._z*other._vector._z))
            vectorPart1=round((self._vector._x*other._a) + (self._a*other._vector._x) + (self._vector._y*other._vector._z) - (self._vector._z*other._vector._y))
            vectorPart2=round((self._a*other._vector._y) + (self._vector._y*other._a) + (self._vector._z*other._vector._x) - (self._vector._x*other._vector._z))
            vectorPart3=round((self._a*other._vector._z) + (self._vector._z*other._a) + (self._vector._x*other._vector._y) - (self._vector._y*other._vector._x))
            return Quaternion(scalerPart,Vector3(vectorPart1,vectorPart2,vectorPart3))             

        else:
            raise Exception("Incompatible type(s) for multiplication with Quaterion")         
    
    def __truediv__(self,factor:float)->"Quaternion":
        """[divides each element ]"""
        try:
            a=round((self._a/factor),Quaternion.prec);
            vect=self._vector/factor
            return Quaternion(a,vect)
        except ZeroDivisionError as _:  
            print("Cannot divide by zero")
            sys.exit("[INPUT ERROR]Exit_Message:ZeroDivisionError")    
    
    def __repr__(self) -> str:
        """[represent the class object]""" 
        return "Quaternion({},{})".format(self._a,self._vector)    

#Object Instantiation
vector=Vector3(12,21,40)
quaternion=Quaternion(3,vector)

File: Quaternion.cpp
#include<iostream>
#include<conio.h>
#include<math.h>
#include<iomanip>
using namespace std;
float epsilon=1e-10;
class Vector3
{
private:
    float x_comp,y_comp,z_comp;
public:
    
    Vector3(float x=0.0,float y=0.0,float z=0.0)
    {set(x,y,z);}
    
    Vector3(const Vector3& v)                       
    { setXcomp(v.getX());
      setYcomp(v.getY());
      setZcomp(v.getZ());
    }
    
    void set(float x,float y,float z)               
    {setXcomp(x);setYcomp(y);setZcomp(z);}
    
    void setXcomp(float x)                           
    {x_comp=x;}

    void setYcomp(float y)
    {y_comp=y;}

    void setZcomp(float z)
    {z_comp=z;}

    float getX() const     
    {return x_comp;} 

    float getY() const
    {return y_comp;}

    float getZ() const
    {return z_comp;}
    
    float Magnitude()
    {float magSquare=pow(getX(),2)+pow(getY(),2)+pow(getZ(),2); return sqrt(magSquare);}

    bool isUnitVector3()  // float should not be compared,we will  
    {return abs(Magnitude()-1.0)<epsilon;}

    Vector3 Conjugate()
    {   
        Vector3 temp(-1*getX(),-1*getY(),-1*getZ());
        return temp;
    }

    Vector3 operator+(const Vector3& v2)
    {
        Vector3 temp;
        temp.setXcomp(getX()+v2.getX()); temp.setYcomp(getY()+v2.getY()); temp.setZcomp(getZ()+v2.getZ());
        return temp;
    }

    Vector3 operator-(const Vector3& v2)
    {
        Vector3 temp;
        temp.setXcomp(getX()-v2.getX()); temp.setYcomp(getY()-v2.getY()); temp.setZcomp(getZ()-v2.getZ());
        return temp;
    }

    Vector3 operator*(const float& factor)  //scaler multiply
    {
        Vector3 temp;
        temp.setXcomp(getX()*factor); temp.setYcomp(getY()*factor); temp.setZcomp(getZ()*factor);
        return temp;
    }

    Vector3 operator/(const float& factor)
    {
        try
        {            
            Vector3 temp;
            temp.setXcomp(getX()/factor); temp.setYcomp(getY()/factor); temp.setZcomp(getZ()/factor);
            return temp;
        }
        catch(const std::exception& e)
        {
            cout<<"Terminating Program [ERROR MESSAGE]: ZERODIVISIONERROR";
            exit(2);
        }

    }

    void showVector3(int precision=2) const
    {cout<<setprecision(precision)<<"Vector3("<<getX()<<","<<getY()<<","<<getZ()<<")"<<endl;}
    
    ~Vector3(){};
};

class Quaternion
{
private:
    float scalerPart;
    Vector3 vectorPart;
public:

    Quaternion()
    {setScaler(0);}

    Quaternion(Vector3 vector,float scaler=0.0)
    { setScaler(scaler); setVector(vector);}

    void setVector(Vector3& v)
    {vectorPart=Vector3(v);}

    void setScaler(float scaler)
    {scalerPart=scaler;}
        
    float getScalerPart()   const
    {return scalerPart;}

    Vector3 getVectorPart() const
    {return vectorPart;}

    float Norm()
    {
        float normSquare=pow(getScalerPart(),2)+pow(getVectorPart().getX(),2)+pow(getVectorPart().getY(),2)+pow(getVectorPart().getZ(),2);
        return sqrt(normSquare);
    }
    Quaternion Versor()
    {
        try
        {
            Quaternion temp;
            Vector3 v=getVectorPart()/Norm();
            temp.setScaler(getScalerPart()/Norm());
            temp.setVector(v);
            return temp;
        }
        catch(const std::exception& e)
        {
            cout<<"Terminating Program [ERROR MESSAGE]: ZERODIVISIONERROR";
            exit(2);
        }    
    }
    bool isUnitQuaternion()
    {return abs(Norm()-1.0)<epsilon;}

    bool isPureQuaternion()
    {return abs(getScalerPart()-0.0)<epsilon;}
    
    Quaternion Conjugate()
    {
        Quaternion temp;
        temp.setScaler(getScalerPart());
        Vector3 vect=getVectorPart().Conjugate(); 
        temp.setVector(vect);
        return temp;
    }
    Quaternion Reciprocal()   // this is also the inverseQuaternion.
    {
        float normSquare=pow(Norm(),2);
        return Conjugate()/normSquare;
    }
    Quaternion operator+(const Quaternion& Q )
    {
        Quaternion temp;
        temp.setScaler(getScalerPart()+Q.getScalerPart());
        Vector3 vect=getVectorPart()+Q.getScalerPart();
        temp.setVector(vect);
        return temp;
    }  
    Quaternion operator-(const Quaternion& Q )
    {
        Quaternion temp;
        temp.setScaler(getScalerPart()-Q.getScalerPart());
        Vector3 vect=getVectorPart()-Q.getScalerPart();
        temp.setVector(vect);
        return temp;
    }
    //For Quaternion multiplication: (i**2= j**2 = k**2 = ijk = -1, ij=k ,jk=i ,ki=j ,ji= -k ,kj= -i ,ik= -j)
    Quaternion operator*(const Quaternion& Q )
    {
        Quaternion temp;
        float scaler   =(getScalerPart()*Q.getScalerPart())        - (getVectorPart().getX()*Q.getVectorPart().getX()) - (getVectorPart().getY()*Q.getVectorPart().getY()) - (getVectorPart().getZ()*Q.getVectorPart().getZ());  
        float vectPart1=(getVectorPart().getX()*Q.getScalerPart()) + (getScalerPart()*Q.getVectorPart().getX())        + (getVectorPart().getY()*Q.getVectorPart().getZ()) - (getVectorPart().getZ()*Q.getVectorPart().getY());     
        float vectPart2=(getScalerPart()*Q.getVectorPart().getY()) + (getVectorPart().getY()*Q.getScalerPart())        + (getVectorPart().getZ()*Q.getVectorPart().getX()) - (getVectorPart().getX()*Q.getVectorPart().getZ());
        float vectPart3=(getScalerPart()*Q.getVectorPart().getZ()) + (getVectorPart().getZ()*Q.getScalerPart())        + (getVectorPart().getX()*Q.getVectorPart().getY()) - (getVectorPart().getY()*Q.getVectorPart().getX());
        Vector3 v(vectPart1,vectPart2,vectPart3);
        temp.setScaler(scaler);
        temp.setVector(v);
        return temp;
    }

    // Scaler multiplication of Quaternion.
    Quaternion operator*(const float& F)
    {
        Quaternion temp;
        temp.setScaler(getScalerPart()*F);
        Vector3 vect=getVectorPart()*F;
        temp.setVector(vect);
        return temp;
    }

    Quaternion operator/(const float& F)
    {
        try
        {
            Quaternion temp;
            temp.setScaler(getScalerPart()/F);
            Vector3 vect=getVectorPart()/F;
            temp.setVector(vect); 
            return temp;
        }
        catch(const std::exception& e)
        {
            cout<<"Terminating Program [ERROR MESSAGE]: ZERODIVISIONERROR";
            exit(2);            
        }
    }
    void showQuaternion(int precision=4)
    {cout<<setprecision(precision)<<"Quaternion("<<getScalerPart()<<","<<"Vector3("<<getVectorPart().getX()<<","<< getVectorPart().getY()<<","<< getVectorPart().getZ()<<"))"<<endl;}
    
    ~Quaternion(){};
};

int main(int argc, char const *argv[])
{
    // Object Instantiation.
    Quaternion q1,q2(Vector3(0,9,2),1);
    Vector3 v2(12,31,11);
    q1.setScaler(12);
    q1.setVector(v2);
    q2.showQuaternion();
    (v2*3).showVector3();
    return 0;
}
